lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1171,6 +1173,112 @@ doesAccessorNeedDynamicAttribute(AccessorDecl *accessor, Evaluator &evaluator) {
   }
 }
 
+llvm::Expected<bool>
+IsFinalRequest::evaluate(Evaluator &evaluator, ValueDecl *decl) const {
+  switch (decl->getKind()) {
+    case DeclKind::Var: {
+      // Properties are final if they are in a static class or a 'let'
+      auto *VD = cast<VarDecl>(decl);
+      auto *PBD = VD->getParentPatternBinding();
+      if (PBD != nullptr) {
+        if (auto *nominalDecl = VD->getDeclContext()->getSelfNominalTypeDecl()) {
+          // If this variable is a class member, mark it final if the
+          // class is final, or if it was declared with 'let'.
+          if (inferFinalAndDiagnoseIfNeeded(decl, PBD->getStaticSpelling()))
+            return true;
+
+          if (VD->isLet() && isa<ClassDecl>(nominalDecl)) {
+
+            if (VD->getFormalAccess() == AccessLevel::Open) {
+              auto &context = decl->getASTContext();
+              auto diagID = diag::implicitly_final_cannot_be_open;
+              if (!context.isSwiftVersionAtLeast(5))
+                diagID = diag::implicitly_final_cannot_be_open_swift4;
+              auto inFlightDiag =
+                context.Diags.diagnose(decl, diagID,
+                                       static_cast<unsigned>(ImplicitlyFinalReason::Let));
+              fixItAccess(inFlightDiag, decl, AccessLevel::Public);
+            }
+
+            return true;
+          }
+        }
+      }
+
+      // Storage is final if its Get accessor is final.
```

Comment:
Doesn't this create a cycle? Below you say "// If the storage is final, propagate to this accessor.". However checking if storage is final will check the getter.

I think the finality of the storage should be the point of truth, with accessors deriving from that.

Reply:
Right, I fixed it to derive the accessors from the storage.